configfile: 'config.yaml'

import os

input_bam = config['bam']
gtf_filename = config['gtf_filename']

intermediate_dir = config['intermediate_dir']
log_dir = config['log_dir']
results_dir = config['results_dir']

class Fasta:
    def __init__(self, ref_path):
        self.fasta = ref_path
        self.fai = self.fasta + '.fai'
        self.references = []
        self.lengths = []
        assert os.path.isfile(self.fasta)
        assert os.path.isfile(self.fai)

        # get chromosome info for 1~22 + X, Y only
        self.process_fai() # only_canonical=True

    def process_fai(self, only_canonical=True):
        for line in open(self.fai):
            field = line.rstrip().split('\t')
            reference, length, cumlen, _a, _b = field
            if only_canonical and (reference == 'MT' or 
                                   reference.startswith('GL')):
                continue
            self.references.append(reference)
            self.lengths.append(int(length))

def generate_intervals(ref, chromosomes, size=1000000): # FROM wgs package; bug workaround
    fasta = Fasta(ref)
    lengths = fasta.lengths
    names = fasta.references

    intervals = []

    for name, length in zip(names, lengths):
        if name not in chromosomes:
            continue
        for i in range(int((length / size) + 1)):
            start = str(int(i * size) + 1)
            end = str(int((i + 1) * size))
            intervals.append(name + "_" + start + "_" + end)

    return intervals

chromosomes = [str(x) for x in range(1, 23)] + ['X', 'Y']
interval_size = 1e7 # 1e9 to cover all chroms
intervals = generate_intervals(config['reference'], chromosomes, interval_size)


rule all:
    input: 
        os.path.join(results_dir, f'roh_qc.pdf'),

rule samtools_germline:
    input: 
        bam=input_bam,
        ref=config['reference']
    output: 
        os.path.join(intermediate_dir, 
            'samtools_germline.{sample}.{interval}.vcf.gz')
    log: 
        os.path.join(log_dir, 
            'samtools_germline.{sample}.{interval}.vcf.gz.log')
    params:
        region = lambda wc: "{}:{}-{}".format(*wc.interval.split('_'))
    singularity: "docker://shahcompbio/roh_qc:0.0.2"
    shell: 
        "samtools mpileup -ugf {input.ref} -Q 20 -q 10 "
        "-r {params.region} {input.bam} 2> {log} | "
        "bcftools call -vmO z -o {output} &>> {log}"

rule bcftools_concat:
    input:
        expand(os.path.join(intermediate_dir, 
            'samtools_germline.{sample}.{interval}.vcf.gz'),
            interval=intervals)
    output:
        os.path.join(intermediate_dir, f'samtools_germline.vcf.gz')
    log: 
        os.path.join(log_dir, f'samtools_germline.vcf.log')
    singularity: "docker://shahcompbio/roh_qc:0.0.2"
    shell: 
        "bcftools concat -O z -o {output} "
        "{input} 2> {log}"

rule bcftools_sort:
    input:
        os.path.join(intermediate_dir, f'samtools_germline.vcf.gz')
    output:
        os.path.join(intermediate_dir, f'samtools_germline.sorted.vcf.gz')
    log: 
        os.path.join(log_dir, f'samtools_germline.sorted.vcf.log')
    singularity: "docker://shahcompbio/roh_qc:0.0.2"
    shell: 
        "bcftools sort -m 4G -O z -o {output} -T tmp {input} 2> {log}"

rule bcftools_roh:
    input:
        os.path.join(intermediate_dir, f'samtools_germline.sorted.vcf.gz')
    output:
        os.path.join(results_dir, f'samtools_germline.roh.tsv')
    log: 
        os.path.join(log_dir, f'samtools_germline.roh.tsv.log')
    singularity: "docker://shahcompbio/roh_qc:0.0.2"
    shell: 
        "bcftools roh -G30 --AF-dflt 0.4 {input} "
        "> {output} 2> {log}"

rule plot_roh:
    input:
        os.path.join(results_dir, f'samtools_germline.roh.tsv')
    output:
        os.path.join(results_dir, f'roh_qc.pdf')
    log: 
        os.path.join(log_dir, f'roh_qc.log')
    singularity: "docker://shahcompbio/roh_qc:0.0.2"
    shell: 
        "python /juno/work/shah/users/chois7/packages/wgs_qc_utils/wgs_qc_utils/roh_qc/plot_roh.py " 
        #"python /usr/bin/plot_roh.py " 
        f"--roh {input} --sample "
        "--pdf {output} &> {log}"

